src/overlays/actors/ovl_Bg_Ddan_Jd/z_bg_ddan_jd.c

```diff
@@ -31,17 +33,157 @@ const ActorInit Bg_Ddan_Jd_InitVars = {
     (ActorFunc)BgDdanJd_Update,
     (ActorFunc)BgDdanJd_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Init.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Destroy.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
+
+#define STATE_GO_BOTTOM 0
+#define STATE_GO_MIDDLE_FROM_BOTTOM 1
+#define STATE_GO_MIDDLE_FROM_TOP 2
+#define STATE_GO_TOP 3
+
+#define MOVE_HEIGHT_MIDDLE 140.0f
+#define MOVE_HEIGHT_TOP 700.0f
+
+#define IDLE_FRAMES 100
+
+// since stepSpeed also determines whether the platform is the shortcut to the top of the dungeon, these must be
+// different values to work correctly
+#define DEFAULT_STEP_SPEED 1
+#define SHORTCUT_STEP_SPEED 5
+
+void BgDdanJd_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+    s32 pad;
+    ColHeader* colHeader = NULL;
+
+    Actor_ProcessInitChain(thisx, &sInitChain);
+    DynaPolyInfo_SetActorMove(this, DPM_PLAYER);
+    DynaPolyInfo_Alloc(&D_06003CE0, &colHeader);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, thisx, colHeader);
+    this->idleTimer = IDLE_FRAMES;
+    this->state = STATE_GO_BOTTOM;
+
+    // @bug: does not check if actor.params < 0x40. This will cause inconsistent behavior if the bound switch flag state
+    // changes when the actor is loaded vs not loaded
+    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params) != 0) {
```

Comment:
I'm not sure I understand the issue here
if it's just that `this->dyna.actor.params` should be `this->dyna.actor.params & 0x3F` I don't think that should be considered a bug, all variables have an implicit format

the `!= 0` can be dropped

Reply:
The bug is a bit complex to explain: 

The implementation of Flags_GetSwitch doesn't prevent a leftshift that is bigger than the wordsize/smaller than 0 from occurring, which I think is  technically undefined behavior in C. On MIPS architecture, the right operand's 5 least significant bits are used to determine how much to left shift by, meaning that there's an implicit mod 32 on the right operand. If you pass in a flag value of 0x20 or higher to Flags_GetSwitch, it will return a temp flag state equal to flag%32. 

Now, to differentiate between the shortcut platform and the normal platform, the game uses the `stepState` variable's state; if stepState is 5, then the actor instance is always treated as the shortcut platform, but if `stepState` is 1, then it checks to see if params < 0x40 to determine if it should become the shortcut platform everywhere else... except here.

So let's say someone decided to spawn an instance where params = 0x0100; this ties the instance to temp switch flag 0x00, and let's say temp switch 0x00 is currently off.  The platform height will alternate from lowest point to middle point.

if you hit a switch or something that changes temp switch 0x00 to on while in the same room, the platform's behavior won't change, but if you leave the room and return, suddenly the init gets run again and now it's going all the way to the top because it doesn't do the range check on params